src/logging.rs

```diff
@@ -1,8 +1,55 @@
 use std::{borrow::Cow, str::FromStr};
 
 use sentry::types::Dsn;
+use serde::{Deserialize, Serialize};
+use tracing::level_filters::LevelFilter;
 
-use crate::config::Config;
+use crate::config;
+
+use tracing_subscriber::{prelude::*, Layer};
+
+/// Controls the log format.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Deserialize, Serialize, clap::ValueEnum)]
+#[serde(rename_all = "lowercase")]
+pub enum LogFormat {
+    /// Auto detect the best format.
+    Auto,
+
+    /// Pretty printing with colors.
+    Pretty,
+
+    /// Simplified plain text output.
+    Simplified,
+
+    /// Dump out JSON lines.
+    Json,
+}
+
+/// The logging level
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, Deserialize, Serialize, clap::ValueEnum)]
+#[serde(rename_all = "lowercase")]
+pub enum Level {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+    Off,
+}
```

Comment:
Not totally in love with the fact that both of these are deriving `clap::ValueEnum` but oh well

Reply:
I would suggest using the `tracing` enum `LevelFilter` instead of rolling your own: https://docs.rs/tracing/latest/tracing/level_filters/struct.LevelFilter.html.

Reply:
Unfortunately I need to derive `clap::ValueEnum`

Reply:
Ah, I see. Type class coherence/the orphan rule continues to be a massive pain.